staging/src/k8s.io/apiserver/pkg/endpoints/filters/constrained_impersonation.go

```diff
@@ -0,0 +1,375 @@
+			continue
+		}
+
+		hasUserExtra = true
+		extraKey := unescapeExtraKey(strings.ToLower(headerName[len(authenticationv1.ImpersonateUserExtraHeaderPrefix):]))
+
+		// make a separate request for each extra value they're trying to set
+		for _, value := range values {
+			attrs := newAttributeRecord(requestor)
+			attrs.Verb = impersonateUserInfoVerb
+			attrs.Name = value
+			attrs.Subresource = extraKey
+			attrs.Resource = "userextras"
+			attrsList = append(attrsList, attrs)
+			userExtra[extraKey] = append(userExtra[extraKey], value)
+		}
+	}
+
+	hasUID := len(uid) > 0
+	if hasUID {
+		attrs := newAttributeRecord(requestor)
+		attrs.Verb = impersonateUserInfoVerb
+		attrs.Resource = "uids"
+		attrs.Name = uid
+		attrsList = append(attrsList, attrs)
+	}
+
+	// If the user is service account and groups is empty, set default service account groups.
+	if namespace, _, err := serviceaccount.SplitUsername(requestedUser); err == nil {
+		if !hasGroups {
+			groups = serviceaccount.MakeGroupNames(namespace)
+		}
+	}
+
+	if (hasGroups || hasUserExtra || hasUID) && !hasUser {
+		return nil, nil, fmt.Errorf("requested %v without impersonating a user", attrsList)
+	}
+
+	if requestedUser != user.Anonymous {
+		// When impersonating a non-anonymous user, include the 'system:authenticated' group
+		// in the impersonated user info:
+		// - if no groups were specified
+		// - if a group has been specified other than 'system:authenticated'
+		//
+		// If 'system:unauthenticated' group has been specified we should not include
+		// the 'system:authenticated' group.
+		addAuthenticated := true
+		for _, group := range groups {
+			if group == user.AllAuthenticated || group == user.AllUnauthenticated {
+				addAuthenticated = false
+				break
+			}
+		}
+
+		if addAuthenticated {
+			groups = append(groups, user.AllAuthenticated)
+		}
+	} else {
+		addUnauthenticated := true
+		for _, group := range groups {
+			if group == user.AllUnauthenticated {
+				addUnauthenticated = false
+				break
+			}
+		}
+
+		if addUnauthenticated {
+			groups = append(groups, user.AllUnauthenticated)
+		}
+	}
+
+	newUser := &user.DefaultInfo{
+		Name:   requestedUser,
+		Groups: groups,
+		Extra:  userExtra,
+		UID:    uid,
+	}
+
+	return attrsList, newUser, nil
+}
+
+func newAttributeRecord(requestor user.Info) *authorizer.AttributesRecord {
+	return &authorizer.AttributesRecord{
+		APIGroup:        authenticationv1.SchemeGroupVersion.Group,
+		APIVersion:      authenticationv1.SchemeGroupVersion.Version,
+		User:            requestor,
+		ResourceRequest: true,
+	}
+}
+
+func authorizeConstrainedImpersonation(ctx context.Context, targetAttributesList []*authorizer.AttributesRecord, a authorizer.Authorizer) (authorizer.Decision, string, authorizer.Attributes, error) {
+	// Get request attributes from the context.
+	requestorAttrs, err := GetAuthorizerAttributes(ctx)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, "", &authorizer.AttributesRecord{}, err
+	}
```

Comment:
move this to point of use below the for loop

Reply:
updated